### PR TITLE
♻️ Refactor build process, CI and documentation

### PR DESCRIPTION
**Pull request by chrisjsewell**
Head-lines:

- `.github/workflows/build.yml` now runs the full vagrant box build, clean, compacting and image creation.
- This PR deprecates https://github.com/marvel-nccr/quantum-mobile-cloud-edition

This PR re-works the build process to be more streamlined:

1. The `yaml_parser.sh`, `compact_hd.sh` and `create_image.sh` scripts have been consolidated into `playbook-package.yml`. This process writes the image and release notes into a `dist/` folder.
2. `playbook.yml` -> `playbook-build.yml`
3. `hosts` and `globalconfig.yml` have been consolidated into `inventory.yml`, and the read `globalconfig.yml` tasks have been removed. This contains all the global variables, and also host specific variables; effectively deprecating the `quantum-mobile-cloud` repository. To run against a specific host, simply call `ansible-playbook playbook-build.yml --extra-vars "build_hosts=aws"`
4. In the `Vagrantfile`, the ansible provisioning now runs against a static IP, which is specified in `inventory.yml`. This allows for ansible to use all the variables defined in `inventory.yml`, rather than working from a separate auto-generated inventory, ensuring consistency.
5. Add a pre-commit for auto-formatting and linting yaml/ansible files which is also run in GH actions via `.github/workflows/ci.yml`
6. A tox.ini has been created from which to run the process. This means you do not have to specifically create a python environment, you simply install [tox](https://tox.readthedocs.io/en/latest/). Then run `tox -e py38-vagrant` (installs all python/ansible role dependencies then runs `vagrant up`), or `tox -e py38-ansible`, when running ansible against an existing VM, or `tox -e py38-package` to run the VM compacting and image creation tasks.
7. The above use of tox, can be seen in `.github/workflows/build.yml`; this runs the full vagrant box build and packaging to completion via GH actions. Obviously this won't be run for every PR commit (as it is currently set up to do for testing), but could be run for each commit to master or something like that. Ideally, this would actually upload the build to somewhere, then we could just manually download/test it before releasing. I was trying to get it to save as an artifact, but currently there is an issue with this (see https://github.com/actions/upload-artifact/issues/29#issuecomment-706512345), we could probably set it up to upload straight to openstack though, or even a google drive account (https://github.com/marketplace/actions/upload-files-to-google-drive). The two current limitations are that 1) you can only set the build VM to have 1 CPU 2) At least with the base installed software on the GH VM, setting `-accelerate3d` to `on` fails the VM start up.

Still TODO:

- [x] update/add documentation
- [x] limit build.yml GH workflow triggers (i.e. not all PR commits)

Other notes (for seperate PRs): 

- currently the virtualenv built is actually broken, because the numpy version is not compatible with aiidalab (this should be fixed in the next release).
- the wannier90 has a clean task, but it fails because it appears to be running against the wrong python executable in the full build, since it can't find the configparser module, pip installed in an earlier task.
- Should add a `release_notes` task to aiidalab, specifying its version
- `aiida_venv` and `aiidalab_venv` should probably be merged
- in `aiida` role, write plugin package versions to separate release_notes section
- improve `resources/Release.md.j2` (add build versions: Virtual Mobile, vagrant, ansible)

@ltalirz @borellim, since it's just the documentation left to do, I would encourage you to give a review.

### CODE REVIEW

**Comment by chrisjsewell** [comment]
>  just for the addition of tox it is not clear to me whether this really makes things easier

trust me, tox makes everything easier 😉

Yeh no problem. FYI I would still like to get auth access to the openstack account, so I can upload myself

**Comment by chrisjsewell** [comment]
> Got the website up and running: [quantum-mobile.readthedocs.io/en/gh-actions](https://quantum-mobile.readthedocs.io/en/gh-actions/) (still needs some work)

the documentation is basically finished, if you want to take a skim @ltalirz

**Review by ltalirz — CHANGES_REQUESTED** [review]
Thanks a lot @chrisjsewell !
It's great to have the build on github actions now; this should already help pinpointing issues and, eventually, might absolve us from having to build locally altogether.
Thanks also for creating the documentation page; over time more and more information accumulated and it makes sense to collect it in one place.

I've had a look through most of the relevant files and added my comments here and there.

As discussed previously, I'm still not convinced of including release information in the docs and I suggest to change this back - we will in any case need to put the release notes on github as well (the github release page has been linked to from many places + if we don't publish releases there it will look like we're not doing releases).
Since the QM cloud and desktop editions are now synchronized in this repo, we can archive the cloud edition repo and release both versions in one go from here, i.e. I don't see a disadvantage in this respect.
Happy to discuss this in more detail.

For the wiki I guess it's ok to include it in the docs (probably it will also change less frequently now than before). Don't forget to put redirect notices from the wiki in place since, again, the FAQs are linked from the Desktop of the QM (as well as from other places).

File: .github/workflows/build.yml
Comment (chrisjsewell) [reply .github/workflows/build.yml:44]: I don't think it will error at all TBH. It was just me being super cautious, not wanting a 1.5 hour build to fail because of a non-essential step

File: .readthedocs.yml
Comment (chrisjsewell) [reply .readthedocs.yml:6]: I felt it was nicer to have all the requirements files next to each other. But I'm not too bothered really, so can change if you want.

File: docs/index.md
Comment (chrisjsewell) [reply docs/index.md:75]: fair

File: playbook-build.yml
Comment (chrisjsewell) [reply playbook-build.yml:92]: No this particular PR does not actually update any roles; the final build should be that of 20.06.1 (so I could check everything was still working against a "known" working set up with this new build process).
Then the next PR, will actually update the roles and be the release.

File: playbook-build.yml
Comment (chrisjsewell) [reply playbook-build.yml:119]: yeh I'm working on this anyway with the in-progress aiida/aiidalab role PRs

File: tox.ini
Comment (chrisjsewell) [reply tox.ini:32]: magic!

No, if you look in the playbook-build.yml, you'll see that the host defaults to `vagrant-ssh`, but will is specified if you run `tox -e py38-ansible -- --extra-vars "build_hosts=aws"`

This is all specified in the documentation

File: resources/Release.md.j2
Comment (chrisjsewell) [reply resources/Release.md.j2:20]: I'd say its not something that has actually changed in the repo or the VM, its just how this particular VM was built.

File: docs/developers/build-cloud.md
Comment (chrisjsewell) [reply docs/developers/build-cloud.md:90]: Why should you need to understand what's going on under the hood?
I think you are just saying this because you are used to running the `ansible-playbook` command, which is understandable.
But IMO this is just an added improvement in ansible's goal for automation and repeatability, in that tox is also automating your python environment setup (plus installing the ansible roles).

File: docs/developers/build-vagrant.md
Comment (chrisjsewell) [reply docs/developers/build-vagrant.md:77]: This is what the original documentation said, so I left it in lol.
But yeh happy to remove.

Comment (chrisjsewell) [inline docs/developers/build-vagrant.md:77]:
```suggestion
# improves interface
```

File: .github/workflows/build.yml
Comment (chrisjsewell) [reply .github/workflows/build.yml:64]: done

File: docs/index.md
Comment (chrisjsewell) [reply docs/index.md:54]: done

File: docs/index.md
Comment (chrisjsewell) [reply docs/index.md:93]: You actually had Max twice in the original README and no Marvel 😛 
added 👍

File: docs/developers/build-vagrant.md
Comment (chrisjsewell) [reply docs/developers/build-vagrant.md:107]: I moved the section, but I'm not sure what to say about modifying the playbook?
My advice would be don't do it, unless you really know what you are doing.
The variables and tags should suffice for most customisations, modifying the playbook is more of a maintainer action.

File: resources/Release.md.j2
Comment (chrisjsewell) [reply resources/Release.md.j2:32]: sure, although if the release instructions are part of the documentation, they don't actually need to be in there at all 😉

File: README.md
Comment (chrisjsewell) [reply README.md:24]: done

**Comment by chrisjsewell** [comment]
Thanks for the review @ltalirz!

I think I've covered all your comments now

> As discussed previously, I'm still not convinced of including release information in the docs and I suggest to change this back

😬 we'll agree to disagree on this. I think it's a horrible user experience, to tell them they have to go find releases on GitHub.
Yes make releases, but then defer the release notes to the repository, where the information can be properly maintained and version controlled.
This is what I do in all my repositories, for example: https://github.com/executablebooks/MyST-Parser/releases, and it what many other prominent repositories do, such as: https://github.com/jupyterlab/jupyterlab/blob/master/docs/source/getting_started/changelog.rst
It is also discouraged by: https://keepachangelog.com/en/1.0.0/#frequently-asked-questions


> Don't forget to put redirect notices from the wiki in place since, again, the FAQs are linked from the Desktop of the QM (as well as from other places).

Yeh I intended to remove the content of the wiki and put links to the documentation.
Also `marvel-nccr.quantum_mobile_customizations/templates/faq.desktop` should be altered to point to the documentation.

File: docs/developers/build-vagrant.md
Comment (chrisjsewell) [reply docs/developers/build-vagrant.md]: done

File: tox.ini
Comment (ltalirz) [reply tox.ini:32]: Thanks - I meant that your comment in the line above says "run against a specific host" but your line doesn't specify a host.
Probably replace `clean=true` by `build_hosts=aws` ?

File: resources/Release.md.j2
Comment (ltalirz) [reply resources/Release.md.j2]: > sure, although if the release instructions are part of the documentation, they don't actually need to be in there at all 😉

One might think that but I'm not sure it is true.
The question is not just whether troubleshooting information (go to FAQ/ mailing list) is available in general but whether people see it when they need it.

P.S. of course I put a wrong link and meant https://github.com/marvel-nccr/quantum-mobile/releases/tag/20.06.1

File: resources/Release.md.j2
Comment (ltalirz) [reply resources/Release.md.j2:26]: Well, changing the base image does change the VM, doesn't it?
I would simply reintroduce the "build process" section, of which this used to be a part: https://github.com/marvel-nccr/quantum-mobile/releases/tag/20.06.1

**Comment by ltalirz** [comment]
Thanks Chris, I've gone through the comments and resolved most of them (+ replied in some).

As for the release notes, how about this: let's indeed keep the detailed changelog in the docs (there was anyhow a [changelog.md](https://github.com/marvel-nccr/quantum-mobile/blob/20.06.1/CHANGELOG.md) in the repository before - just checking that this information is not lost?), and link to it from the github releases page but let's keep the installation instructions and contact info in the release notes (which you anyhow don't want to replicate in a changelog).

The changelog in the docs could in principle be just on a single page (easier to scan/search) but if you prefer individual posts you can of course do that as well.


I.e. the release notes on github would be


## Installation instructions

Get Quantum Mobile running on your computer in three simple steps:

 1. Download virtual machine image (3.4G)
    https://bit.ly/3gEZqDL
    Filename: `quantum_mobile_20.06.1.ova`
    MD5 hash: `dba6be9e60e5adf1618a6e251e579c1b`
 2. Install Virtual Box 6.1.6 or later
    www.virtualbox.org
 3. Import virtual machine image into Virtualbox (9.9G)
    File => Import Appliance

Login credentials: username: `max`, password: `moritz`.
The default configuration of `2` cores and `1536` MB RAM can be adjusted in the VM settings.

## Contact

For issues encountered during installation, please first consult the [FAQ page](https://github.com/marvel-nccr/quantum-mobile/wiki/Frequently-Asked-Questions#virtualbox-installationstartup-issues).

Please direct inquiries regarding Quantum Mobile to the [AiiDA mailinglist](http://www.aiida.net/mailing-list/)

## Changelog

See the [Quantum Mobile documentation]().

File: docs/developers/build-vagrant.md
Comment (ltalirz) [inline docs/developers/build-vagrant.md:5]:
```suggestion
The [quantum mobile repository](https://github.com/marvel-nccr/quantum-mobile) contains the Vagrant and [Ansible playbooks](https://docs.ansible.com/ansible/latest/user_guide/playbooks.html) required to set up the virtual machine from scratch.
```

File: docs/developers/build-vagrant.md
Comment (ltalirz) [inline docs/developers/build-vagrant.md:35]:
```suggestion
See the [continuous deployment (CD) workflow](https://github.com/marvel-nccr/quantum-mobile/actions?query=workflow%3ACD) for up-to-date timings
```

File: docs/developers/build-vagrant.md
Comment (ltalirz) [inline docs/developers/build-vagrant.md:43]: I guess this should be logically at the top (I added it later)

File: docs/developers/build-vagrant.md
Comment (ltalirz) [inline docs/developers/build-vagrant.md:58]: Note to self: I'll make a small PR once this is merged, to harmonize this with the "Create the virtual machine" and "customizing QM" sections

**Comment by ltalirz** [comment]
One question: since we're anyhow using tox to create the environment, is there a need to support python 3.6-3.8?
```
$ tox -a -v
using tox.ini: /Users/leopold/Personal/Postdoc-MARVEL/repos/quantum-mobile/tox.ini (pid 23987)
using tox-3.20.1 from /Users/leopold/Applications/miniconda3/envs/qmobile/lib/python2.7/site-packages/tox/__init__.pyc (pid 23987)
default environments:
py38-vagrant -> Build a vagrant box VM then provision with ansible

additional environments:
py37-vagrant -> Build a vagrant box VM then provision with ansible
py36-vagrant -> Build a vagrant box VM then provision with ansible
py36-ansible -> Run ansible on an existing VM
py38-package -> Package a built Vagrant box for distribution
docs-clean   -> Build the documentation
py37-ansible -> Run ansible on an existing VM
py37-package -> Package a built Vagrant box for distribution
py36-package -> Package a built Vagrant box for distribution
docs-update  -> Build the documentation
py38-ansible -> Run ansible on an existing VM
docs-live    -> Build the documentation and launch browser
```
Is there any downside to just offer one (in which case one can also drop the py.. prefix)?

File: Vagrantfile
Comment (ltalirz) [inline Vagrantfile:89]: since this will be the main output people see, I suggest we comment this (in most cases, verbose output should not be needed)

Comment (chrisjsewell) [reply Vagrantfile:89]: yeh makes sense 👍

**Comment by chrisjsewell** [comment]
or it look like you can just use `py3-`: https://tox.readthedocs.io/en/latest/config.html#conf-basepython

File: tox.ini
Comment (ltalirz) [reply tox.ini:32]: @chrisjsewell ping 1

File: resources/Release.md.j2
Comment (ltalirz) [reply resources/Release.md.j2:26]: @chrisjsewell ping 2

**Comment by ltalirz** [comment]
@chrisjsewell I've run the commands and it seems to have worked just fine :-) (I also didn't test the VM; but that's for later)

I've pinged you in two open comments from the first review; apart from that there are just the suggestions from above and we should be good to go.

File: Vagrantfile
Comment (chrisjsewell) [inline Vagrantfile:89]:
```suggestion
    # ansible.verbose = "v"
```

File: docs/developers/build-vagrant.md
Comment (chrisjsewell) [inline docs/developers/build-vagrant.md:43]:
```suggestion
- Host OS: Building Quantum Mobile has been tested on MacOS, Ubuntu and Windows.
- [Vagrant](https://www.vagrantup.com/downloads.html) >= 2.0.1
- [VirtualBox](https://www.virtualbox.org/wiki/Downloads) >= 6.1.6
- [Python](https://www.python.org/) >= 3.6
```

File: docs/developers/build-vagrant.md
Comment (chrisjsewell) [reply docs/developers/build-vagrant.md]: done

File: tox.ini
Comment (chrisjsewell) [reply tox.ini]: oh yeh haha 🤦 fixed

**Comment by chrisjsewell** [comment]
@ltalirz in 31547ff I improved the tox (and updated the documentation) so now you just get:

```console
$ tox -a -v
using tox.ini: /Users/chrisjsewell/Documents/GitHub/quantum-mobile/tox.ini (pid 83422)
using tox-3.20.0 from /Users/chrisjsewell/opt/miniconda3/lib/python3.7/site-packages/tox/__init__.py (pid 83422)
default environments:
vagrant     -> Build a vagrant box VM then provision with ansible

additional environments:
docs-update -> Build the documentation (modify any existing build)
docs-clean  -> Build the documentation (remove any existing build)
ansible     -> Run ansible on an existing VM
package     -> Package a built Vagrant box for distribution
docs-live   -> Build the documentation and launch browser
```

BTW have you tried `tox -e docs-live`?
As you modify the documentation, it automatically triggers rebuilds and reloads the browser, which I think is pretty neat 😄

**Review by ltalirz — APPROVED** [review]
@chrisjsewell Great, thanks for all the work, from my side this is good to squash & merge :-)

**Comment by ltalirz** [comment]
> BTW have you tried tox -e docs-live?

I just tried but for me building the `watchdog` wheel failed
```
Command "/Users/leopold/Personal/Postdoc-MARVEL/repos/quantum-mobile-tox/.tox/docs-live/bin/python -u -c "import setuptools, tokenize;__file__='/private/var/folders/8y/7c7qkxcx5vxb749rt3sw5sh40000gp/T/pip-install-n0l20bmr/watchdog/setup.py';f=getattr(tokenize, 'open', open)(__file__);code=f.read().replace('\r\n', '\n');f.close();exec(compile(code, __file__, 'exec'))" install --record /private/var/folders/8y/7c7qkxcx5vxb749rt3sw5sh40000gp/T/pip-record-emrih3p5/install-record.txt --single-version-externally-managed --compile --install-headers /Users/leopold/Personal/Postdoc-MARVEL/repos/quantum-mobile-tox/.tox/docs-live/include/site/python3.6/watchdog" failed with error code 1 in /private/var/folders/8y/7c7qkxcx5vxb749rt3sw5sh40000gp/T/pip-install-n0l20bmr/watchdog/
```

**Comment by chrisjsewell** [comment]
> Great, thanks for all the work, from my side this is good to squash & merge :-)

thanks!

I just made one last change in fec5d9f, to improve the auto-generated release notes, if you want to look

> I just tried but for me building the watchdog wheel failed

oh that's annoying

**Comment by chrisjsewell** [comment]
Also initiated one last test build: https://github.com/marvel-nccr/quantum-mobile/runs/1258369990

**Comment by ltalirz** [comment]
> I just made one last change in fec5d9f, to improve the auto-generated release notes, if you want to look

Thanks, looks good and useful!

> Also initiated one last test build: https://github.com/marvel-nccr/quantum-mobile/runs/1258369990

Ok! Feel free to merge afterwards; I'll then add [this](https://github.com/marvel-nccr/quantum-mobile/pull/131#discussion_r504781762) minor update

File: .github/workflows/build-init.yml
Comment (chrisjsewell) [inline .github/workflows/build-init.yml:5]: @ltalirz I've added this build job to the CI, just to check the build initiation.
It takes about 5 minutest to run, and shouldn't run if only docs are being changed.

File: docs/maintainers/develop.md
Comment (chrisjsewell) [inline docs/maintainers/develop.md:49]: @ltalirz I've changed this to be more in-line with your suggestion in https://github.com/marvel-nccr/quantum-mobile/pull/131#issuecomment-708280513

File: .github/workflows/build-init.yml
Comment (ltalirz) [reply .github/workflows/build-init.yml:5]: sounds good

File: docs/maintainers/develop.md
Comment (ltalirz) [reply docs/maintainers/develop.md:49]: thanks!

**Comment by ltalirz** [comment]
Let's get this PR merged, and then proceed with the inclusion of bigdft and the actual release

**Pull request merged**
